letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +75,123 @@ def validate_domain(self, domain):
+                    "or local domains!"))
 
     @api.model
     def cron(self):
         domain = urlparse.urlparse(
             self.env['ir.config_parameter'].get_param(
                 'web.base.url', 'localhost')).netloc
         self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
-                self.env['ir.config_parameter'].get_param(
-                    'letsencrypt.chain_certificate_address',
-                    'https://letsencrypt.org/certs/'
-                    'lets-encrypt-x3-cross-signed.pem')
-            )
-            crt.write(chain_cert.read())
-            chain_cert.close()
-            _logger.info('wrote %s', crt.name)
-        reload_cmd = self.env['ir.config_parameter'].get_param(
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+        # Generate account key
+        account_key_file = self._generate_key('account_key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        client = self._create_client(account_key)
+        new_reg = NewRegistration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        client.new_account(new_reg)
+        csr = self._make_csr(account_key, domain_key_file, domain)
+        authzr = client.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        # we don't care what's in the CSR, domains,
+        # whatever challenges letsencrypt sends back, respond
+        # prefer http though.
+        for authorizations in authzr.authorizations:
+            for challenge in authorizations.body.challenges:
+                # we might be given multiple challenges,
+                # we only need to respond to one.
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP:
+                    self._respond_challenge_http(challenge, account_key)
+                    break
+                elif challenge.chall.typ == TYPE_CHALLENGE_DNS:
+                    self._respond_challenge_dns(challenge, domain)
```

Comment:
we will care very much for which domain this is a challenge, how are you going to handle requesting a cert for `*.therp.eu` and `*.therp.nl` together here?